Preserve order of variables in combine_by_coords

kmuehlbauer opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #8828
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

kmuehlbauer commented:
I think this was from times where we had to deal with unsorted dict. At least that was what I understood from a previous comment at that code position.

kmuehlbauer commented:
> Generally, the iterable needs to already be sorted on the same key function.

Now, in light of this... But somehow it works.

kmuehlbauer commented:
OK, let me add some more testing to be sure this works in datasets of any order.

keewis commented:
the reason `itertoolsgroupby` needs sorted iterables is that it combines groups locally, not globally. So `list(itertools.groupby("aabac", key=lambda x: x))` would result in `["a", "b", "a", "c"]`, not `["a", "b", "c"]` (this is not the case for `more_itertools.bucket` and `toolz.itertoolz.groupby`).

kmuehlbauer commented:
Thanks @keewis. AFAICT, the sorting before groupby makes sure that all Datasets with same variables are grouped together regardless of the variable order within each Dataset. 

Removing the sorting will result in more groups but doesn't break the tests. Does that mean we are undertesting?. Or is it just fixed by the subsequent merge?

The issue which should be solved by this PR is that this sorting rearranges the input objects and when using `compat="override"` might move the first object to another position, resulting in wrong output. Then, I was thinking to special case `compat="override"` to keep the first object at it's place, but didn't find a way to incorporate that with the groupby.

kmuehlbauer commented:
After reading on itertools and collections I've found that it's possible to change itertools.groupby with a collections.defaultdict implementation to preserve order and with some intermediate performance gain:

```python
import xarray as xr
import collections
import itertools

def vars_as_keys(ds):
    return tuple(sorted(ds))

def groupby_defaultdict(iter, key=lambda x: x):
    idx = collections.defaultdict(list)
    for i, obj in enumerate(iter):
        idx[key(obj)].append(i)
    for k, ix in idx.items():
        yield k, (iter[i] for i in ix)
        
def groupby_itertools(iter, key=lambda x: x):
    iter = sorted(iter, key=vars_as_keys)
    return itertools.groupby(iter, key=vars_as_keys)

x1 = xr.Dataset({"a": (("y", "x"), [[1]]),
                 "c": (("y", "x"), [[1]]),
                 "b": (("y", "x"), [[1]]),}, 
                coords={"y": [0], "x": [0]})
x2 = xr.Dataset({"d": (("y", "x"), [[1]]),
                 "a": (("y", "x"), [[2]]),},
                coords={"y": [0], "x": [0]})
x3 = xr.Dataset({"a": (("y", "x"), [[3]]),
                 "d": (("y", "x"), [[2]]),},
                coords={"y": [0], "x": [1]})
data_objects = [x2, x1, x3]
```

```python
%%timeit
grouped_by_vars = groupby_defaultdict(data_objects, key=vars_as_keys)
274 ns ± 0.934 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
```python
%%timeit
grouped_by_vars = groupby_itertools(data_objects, key=vars_as_keys)
4.98 µs ± 14 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

headtr1ck commented on xarray/core/combine.py:
```diff
@@ -606,6 +606,15 @@ def vars_as_keys(ds):
     return tuple(sorted(ds))
 
 
+def groupby_defaultdict(iter, key=lambda x: x):
```
```suggestion
T = TypeVar("T")
K = TypeVar("K", bound=Hashable)

def groupby_defaultdict(iter: list[T], key: Callable[[T], K] = lambda x: x) -> Iterator[tuple[K, Iterator[T]]]:
```
Could you add some type hints here?
Not sure if my suggestion is correct.

kmuehlbauer replied:
Thanks @headtr1ck, but I have no idea what those `TypeVar` declarations are doing. The only thing I know is, that `iter` is a list of `xr.Dataset` and `key` is a function which consumes and returns `xr.Dataset`.

kmuehlbauer replied:
Ah sorry,the key-function returns `tuple(sorted(xr.Datatset))`

kmuehlbauer replied:
That's a tuple of strings, AFAICT.

headtr1ck replied:
Ok, then it makes sense. The return value has to be hashable.

The TypeVars can be anything given the bounds (so K has to be hashable) but it has to be consistent in a function call.

E.g. the key argument needs to be the same type as the list contents.

kmuehlbauer replied:
So, your suggestion will work then? Let's try.

headtr1ck replied:
Oh, you need some imports from typing

headtr1ck commented on xarray/tests/test_combine.py:
```diff
@@ -1043,6 +1043,20 @@ def test_combine_by_coords_incomplete_hypercube(self):
         with pytest.raises(ValueError):
             combine_by_coords([x1, x2, x3], fill_value=None)
 
+    def test_combine_by_coords_override_order(self):
```
```suggestion
    def test_combine_by_coords_override_order(self) -> None:
```

headtr1ck reviewed (approved):
LGTM, some type hints works be nice!

kmuehlbauer commented:
From all what I read about the comparison `itertools.groupby` vs. `defaultdict(list)` both have their use cases. 

Both have `O(n)` time complexity, but the sorting step of `itertools.groupby` makes this a `O(n log n)`. The major PRO for `itertools.groupby` is the lazy yielding of groups, the major CON is the need of pre-sorting (which is the root-cause of #8828 and what this PR should fix).

The major PRO for `defaultdict(list)` is that it can work on unsorted input in a fast way, the major CON is it is storing the whole result in memory.

I'm not sure if there is another solution including `itertools.groupby` to fix #8828, so if someone has another idea of tackling that, please shoot.

kmuehlbauer commented on xarray/core/combine.py:
```diff
@@ -1,8 +1,8 @@
-from collections import Counter
-from collections.abc import Iterable, Iterator, Sequence
+from collections import Counter, defaultdict
+from collections.abc import Callable, Iterable, Iterator, Sequence
```
```suggestion
from collections.abc import Callable, Hashable, Iterable, Iterator, Sequence
```

kmuehlbauer commented on xarray/core/combine.py:
```diff
@@ -606,6 +606,21 @@ def vars_as_keys(ds):
     return tuple(sorted(ds))
 
 
+T = TypeVar("T")
```
```suggestion
```

kmuehlbauer commented on xarray/core/combine.py:
```diff
@@ -606,6 +606,21 @@ def vars_as_keys(ds):
 
 
+T = TypeVar("T")
+K = TypeVar("K", bound=Hashable)
+
+
+def groupby_defaultdict(
+    iter: list[T], key: Callable[[T], K] = lambda x: x
+) -> Iterator[tuple[K, Iterator[T]]]:
+    """replacement for itertools.groupby"""
+    idx = defaultdict(list)
+    for i, obj in enumerate(iter):
+        idx[key(obj)].append(i)
+    for k, ix in idx.items():
+        yield k, (iter[i] for i in ix)
```
@headtr1ck 

Can you make anything out of this:

```
xarray/core/combine.py: note: In function "groupby_defaultdict":
xarray/core/combine.py:614: error: Incompatible default for argument "key" (default has type "Callable[[T], T]", argument has type "Callable[[T], K]")  [assignment]
xarray/core/combine.py:614: error: Incompatible return value type (got "T", expected "K")  [return-value]
```

headtr1ck replied:
Interesting.

I guess the issue is when you pass a list of non-hashables (like the list of Datasets) and use the default, which ofc would not work.

I would have guessed that the type checker would complain when calling it like that, not in the function definition....

Not sure if that is a mypy bug or a feature. 
Have to think about it more.

kmuehlbauer replied:
@headtr1ck Coming back to this, let's just assume `key` as positional argument.

kmuehlbauer commented:
I've fixed the typing issue and replaced the second occurrence here. For the above shown [minibenchmark](https://github.com/pydata/xarray/pull/9070#issuecomment-2151965535) this a slight performance gain. The asv bench doesn't show a performance chain.

The major win is that it fixes #8828. Ready for another review.

kmuehlbauer commented:
Do you see any blockers here @keewis, @TomNicholas?

kmuehlbauer commented:
I've added this to #10002 and will merge this tomorrow, if there are no complaints.

TomNicholas commented:
Sorry for taking so long to reply to this 😓 

> AFAICT, the sorting before groupby makes sure that all Datasets with same variables are grouped together regardless of the variable order within each Dataset. Update: And the position in the object list.

IIRC the point of the groupby is to group datasets into sets whose variables have _common  concatenation dimensions_. See https://github.com/CWorthy-ocean/roms-tools/issues/223#issuecomment-2611052848 for an example of a use case.

> Removing the sorting will result in more groups but doesn't break the tests. Does that mean we are undertesting?. Or is it just fixed by the subsequent merge?

I _think_ as long as those are just subgroups of the original groups (i.e. those created by the code that does sorting) then yes the subsequent merge should fix it. The reason being that any subset of a group containing variables with common concatenation dimensions should also have those same common concatenation dimensions?

kmuehlbauer commented:
> Sorry for taking so long to reply to this 😓

No worries :-) . So, essentially the proposed code does the same groupby, but keeps the initial order. Did you had a chance to test your notebook against this PR?

kmuehlbauer commented:
Just for the record, the example from above with the according output. We can see that the defaultdict_groupby generates the same groups and keeps them in their natural order. I think this is along the lines the behaviour @keewis mentioned for `more_itertools.bucket` and `toolz.itertoolz.groupby`. So I'd say we can replace the current implementation with the proposed.


```python
import xarray as xr
import collections
import itertools

def vars_as_keys(ds):
    return tuple(sorted(ds))

def groupby_defaultdict(iter, key=lambda x: x):
    idx = collections.defaultdict(list)
    for i, obj in enumerate(iter):
        idx[key(obj)].append(i)
    for k, ix in idx.items():
        yield k, (iter[i] for i in ix)
        
def groupby_itertools(iter, key=lambda x: x):
    iter = sorted(iter, key=vars_as_keys)
    return itertools.groupby(iter, key=vars_as_keys)

x1 = xr.Dataset({"a": (("y", "x"), [[1]]),
                 "c": (("y", "x"), [[1]]),
                 "b": (("y", "x"), [[1]]),}, 
                coords={"y": [0], "x": [0]})
x2 = xr.Dataset({"d": (("y", "x"), [[1]]),
                 "a": (("y", "x"), [[2]]),},
                coords={"y": [0], "x": [0]})
x3 = xr.Dataset({"a": (("y", "x"), [[3]]),
                 "d": (("y", "x"), [[2]]),},
                coords={"y": [0], "x": [1]})
data_objects = [x2, x1, x3]

for vars, ds_with_same_vars in groupby_defaultdict(data_objects, key=vars_as_keys):
    print(vars, tuple(ds_with_same_vars))
print("---------------------------------------------------------------------")
for vars, ds_with_same_vars in groupby_itertools(data_objects, key=vars_as_keys):
    print(vars, tuple(ds_with_same_vars))
```

```python
('a', 'd') (<xarray.Dataset> Size: 32B
Dimensions:  (y: 1, x: 1)
Coordinates:
  * y        (y) int64 8B 0
  * x        (x) int64 8B 0
Data variables:
    d        (y, x) int64 8B 1
    a        (y, x) int64 8B 2, <xarray.Dataset> Size: 32B
Dimensions:  (y: 1, x: 1)
Coordinates:
  * y        (y) int64 8B 0
  * x        (x) int64 8B 1
Data variables:
    a        (y, x) int64 8B 3
    d        (y, x) int64 8B 2)
('a', 'b', 'c') (<xarray.Dataset> Size: 40B
Dimensions:  (y: 1, x: 1)
Coordinates:
  * y        (y) int64 8B 0
  * x        (x) int64 8B 0
Data variables:
    a        (y, x) int64 8B 1
    c        (y, x) int64 8B 1
    b        (y, x) int64 8B 1,)
---------------------------------------------------------------------
('a', 'b', 'c') (<xarray.Dataset> Size: 40B
Dimensions:  (y: 1, x: 1)
Coordinates:
  * y        (y) int64 8B 0
  * x        (x) int64 8B 0
Data variables:
    a        (y, x) int64 8B 1
    c        (y, x) int64 8B 1
    b        (y, x) int64 8B 1,)
('a', 'd') (<xarray.Dataset> Size: 32B
Dimensions:  (y: 1, x: 1)
Coordinates:
  * y        (y) int64 8B 0
  * x        (x) int64 8B 0
Data variables:
    d        (y, x) int64 8B 1
    a        (y, x) int64 8B 2, <xarray.Dataset> Size: 32B
Dimensions:  (y: 1, x: 1)
Coordinates:
  * y        (y) int64 8B 0
  * x        (x) int64 8B 1
Data variables:
    a        (y, x) int64 8B 3
    d        (y, x) int64 8B 2)

```

TomNicholas reviewed (approved):
> We can see that the `groupby_defaultdict` generates the same groups and keeps them in their natural order

This sounds fine to me.
